fix: path parameters are not encoded #679

Maxim-Mazurok opened the pull request:
## Status

<!--- **READY/WIP/HOLD** --->

**READY**

## Description

Addresses https://github.com/anymaniax/orval/issues/679 issue, based on my patch https://github.com/anymaniax/orval/issues/679#issuecomment-1583652672

Also updates vue to the latest, so that we can use `MaybeRef` from it's indirect dependency `@vue/reactivity`. I need this because the old import was unresolved when using `"moduleResolution": "bundler"` which is the default now for vue in in `node_modules/@vue/tsconfig/tsconfig.json`

I think some people might have implemented encoding on their end, so this probably should be a major release, otherwise people will get double-encoding.

## Todos

- [x] Tests
- [x] Documentation
- [ ] Changelog Entry (unreleased) - not sure what to do about this

anymaniax commented:
@Maxim-Mazurok thanks for the help. It's encode every parameters? If wouldn't be better to add that as a config?

Maxim-Mazurok commented:
> @Maxim-Mazurok thanks for the help. It's encode every parameters? If wouldn't be better to add that as a config?

I would argue that it should be encoded by default, for example `/api?song=Me&you` will unexpectedly be parsed as `song=Me; you=`

But I guess it'd be easier to make a non-default option, then no need for major release, I'll do that.

melloware commented:
I would think we want all parameters encoded by default no?

Maxim-Mazurok commented:
> I would think we want all parameters encoded by default no?

I would want that as well, but as @anymaniax mentioned this will be a breaking change and might require a major release. I think that would be a good idea overall, what do you think?

melloware commented:
I mean I guess I don't see the harm in encoding URLs it feels like the safe thing to do. I wonder what @anymaniax concern is?

Maxim-Mazurok commented:
> I mean I guess I don't see the harm in encoding URLs it feels like the safe thing to do. I wonder what @anymaniax concern is?

Well, I'm pretty sure that we agree that it's a good thing, but changing default behaviour will be a breaking change.

For example, if someone had `/api?song=Me&you` in their project - their server would receive `{song: "Me", you: undefined}`. Even thought this is a bug, some people might be fine with this bug and actually want it to work this way. If we make the change and encode all by default - their server will now get `{song: "Me & you"}` which is technically correct, but imagine if they already have some logic like `if (song === "Me")` to handle this request - then it'll break, potentially introducing bugs.

So I think the best thing to do is to encode all params by default and make a major release (major version bump) so that people know to expect breaking changes.

The second best thing to do is to make encoding params a non-default option.

Maxim-Mazurok commented on packages/query/src/index.ts:
```diff
@@ -368,13 +368,22 @@ const generateQueryRequestFunction = (
   }
+
+  if (output?.urlEncodeParameters) {
+    route = makeRouteSafe(route);
```
This is the core change that addresses the original issue

Maxim-Mazurok commented:
This is ready for review now

Summary:
- Added `urlEncodeParameters` option, `false` by default
- Enable `tsc --noEmit` for `packages/orval` (ignored most of the pre-existing issues for now)
- Added docs for `urlEncodeParameters` and `allParamsOptional` new output options
- Added `zod` option for `OutputClient`
- Passing `output: NormalizedOutputOptions` all the way into `packages/query`'s `builder` (now I realise that it probably makes sense to have `urlEncodeParameters` for all clients and I could probably have done it somewhere else... perhaps we can refactor in the future, still I think it's useful to pass config options to the builder)

anymaniax reviewed (approved):
Seems really good to me! Thanks @Maxim-Mazurok

lukaw3d commented on packages/query/src/utils.ts:
```diff
@@ -109,5 +109,14 @@ export const vueUnRefParams = (props: GetterProps): string => {
+  route: string,
+  prepend: string,
+  append: string,
+): string => route.replaceAll(/\${(.+?)}/g, `\${${prepend}$1${append}}`);
```
getRouteAsArray uses a slightly different regex

Maxim-Mazurok replied:
Ah, interesting. I think mine is better because it seems to be more aligned with OpenAPI specs by allowing special characters and an empty string: https://github.com/OAI/OpenAPI-Specification/issues/1969#issuecomment-509016671

But yeah, it's probably better to refactor it to avoid duplication

Maxim-Mazurok replied:
I've created https://github.com/anymaniax/orval/issues/1084 to address that
